pkg/quarterdeck/db/models/users.go

```diff
@@ -302,6 +303,89 @@ func (u *User) UpdateLastLogin(ctx context.Context) (err error) {
 	return tx.Commit()
 }
 
+const (
+	listUserIdSQL = "SELECT user_id FROM organization_users where organization_id = $1"
+	getUserSQL    = "SELECT id, name, email, terms_agreement, privacy_agreement, last_login, created, modified from users where id = $1"
+)
+
+func ListUsers(ctx context.Context, orgID any, prevPage *pagination.Cursor) (users []*User, cursor *pagination.Cursor, err error) {
+	var userOrg ulid.ULID
+	if userOrg, err = ulids.Parse(orgID); err != nil {
+		return nil, nil, err
+	}
+
+	if ulids.IsZero(userOrg) {
+		return nil, nil, ErrMissingOrgID
+	}
+
+	if prevPage == nil {
+		// Create a default cursor, e.g. the previous page was nothing
+		prevPage = pagination.New("", "", 0)
+	}
+
+	if prevPage.PageSize <= 0 {
+		return nil, nil, invalid(ErrMissingPageSize)
+	}
+
+	var tx *sql.Tx
+	if tx, err = db.BeginTx(ctx, &sql.TxOptions{ReadOnly: true}); err != nil {
+		return nil, nil, err
+	}
+	defer tx.Rollback()
+
+	// Fetch list of user_ids associated with the orgID
+	var rows *sql.Rows
+	if rows, err = tx.Query(listUserIdSQL, userOrg); err != nil {
+		if errors.Is(err, sql.ErrNoRows) {
+			return nil, nil, nil
+		}
+		return nil, nil, err
+	}
+	// Process rows into a result page
+	nRows := int32(0)
+	users = make([]*User, 0, prevPage.PageSize)
+	for rows.Next() {
+		// The query will request one additional message past the page size to check if
+		// there is a next page. We should not process any messages after the page size.
+		nRows++
+		if nRows > prevPage.PageSize {
+			continue
+		}
+
+		//create user object to append to the users list and add the orgID to it
+		user := &User{orgID: userOrg}
+		var userID ulid.ULID
+		if err = rows.Scan(&userID); err != nil {
+			return nil, nil, err
+		}
+		if err = tx.QueryRow(getUserSQL, userID).Scan(&user.ID, &user.Name, &user.Email, &user.AgreeToS, &user.AgreePrivacy, &user.LastLogin, &user.Created, &user.Modified); err != nil {
+			if errors.Is(err, sql.ErrNoRows) {
+				return nil, nil, nil
+			}
+			return nil, nil, err
```

Comment:
I'm not super familiar with the underlying database schema, but I'm curious if you think this could be done within a single query (e.g. by using a join?). I don't think that it's necessary to change anything in this PR, so just wondering! If you think it might be possible to refactor this in the future to reduce queries (which could speed up the Raft consensus process, especially if we have orgs with lots of users), maybe you could add a little inline comment here with a TODO so we don't forget? Otherwise feel free to ignore this 😁

Reply:
That's a very good and you are correct, there is a more efficient way to do this.  I decided to refactor the code because it's not a big lift.